hangman/lib/hangman/game.ex

```diff
@@ -0,0 +1,66 @@
+defmodule Hangman.Game do
+
+  defstruct(
+    secret:      [],
+    word_chars:  MapSet.new(),
+    used_chars:  MapSet.new(),
+    game_state:  :initializing,
+    turns_left:  7,
+    letters:     [],
+    last_guess:  ""
+  )
+
+  def new_game() do
+    secret  = Dictionary.random_word() |> String.codepoints
+    %Hangman.Game{
+      secret:     secret,
+      word_chars: MapSet.new(secret),
+      letters:    String.duplicate("_", length(secret)) |> String.codepoints
+    }
+  end
+
+  def tally(game) do
+    %{
+      game_state: game.game_state,
+      turns_left: game.turns_left,
+      letters:    game.letters,
+      used:       game.used_chars |> MapSet.to_list(),
+      last_guess: game.last_guess
+    }
+  end
+
+  #               won?   old?   good?  turns_left
+  defp next_state(true,  false, true,  turns), do: {:won, turns}
+  defp next_state(_____, true,  ____,  turns), do: {:already_used, turns}
+  defp next_state(false, false, true,  turns), do: {:good_guess, turns}
+  defp next_state(false, false, false, 1),     do: {:lost, 0}
+  defp next_state(false, false, false, turns), do: {:bad_guess, turns-1}
+  
+  defp fill_in(guess, [_h1 | t1], [guess | t2]), do: [guess | fill_in(guess, t1, t2)]
+  defp fill_in(guess, [h1 | t1], [_h2 | t2]),    do: [h1 | fill_in(guess, t1, t2)]
+  defp fill_in(_guess, [], []), do: []
+
+  def make_move(%Hangman.Game{game_state: :won} = game, _),  do: {game, tally(game)}
```

Comment:
I'm not deducting anything here, because once you've adopted the decision table approach, your hands are tied. However, in general, I get nervous when I see the entire state recreated on each call, as it means this one function is doing a lot.

Instead, I like to see a transformation of the state as we flow through the function. Each step in the pipeline updates just one or two things before passing the state on.